Compiler/src/Transform.hs

```diff
@@ -0,0 +1,159 @@
+(>>>) :: (ProgramTransform p1 p2, PhaseTransition p1 p1) => (a, Env p1) -> (a -> State (Env p2) b) -> (b, Env p2)
+(a, s) >>> f = runState (f a) (transformEnv s)
+
+-- AST transforms
+class (Phase a, Phase b) => ProgramTransform a b where
+    transformXType :: (Phase c, PhaseTransition a c) => XType a -> State (Env c) (XType b)
+
+transformBaseType :: forall a b c. (Phase a, Phase b, Phase c, ProgramTransform a b, PhaseTransition a c) => BaseType a -> State (Env c) (BaseType b)
+transformBaseType Nat = pure Nat
+transformBaseType PsaBool = pure PsaBool
+transformBaseType PsaString = pure PsaString
+transformBaseType Address = pure Address
+transformBaseType (Named name) = pure $ Named name
+transformBaseType Bot = pure Bot
+
+transformBaseType (Record keys fields) = do
+    transformedFields <- mapM transformVarDef fields
+    pure $ Record keys transformedFields
+
+transformBaseType (Table keys t) = do
+    transformedT <- transformXType @a @b @c t
+    pure $ Table keys transformedT
+
+transformQuantifiedType :: forall a b c. (Phase a, Phase b, Phase c, ProgramTransform a b, PhaseTransition a c) => QuantifiedType a -> State (Env c) (QuantifiedType b)
+transformQuantifiedType (q, t) = do
+    transformedT <- transformBaseType t
+    pure (q, transformedT)
+
+transformVarDef :: forall a b c. (Phase a, Phase b, Phase c, ProgramTransform a b, PhaseTransition a c) => VarDef a -> State (Env c) (VarDef b)
+transformVarDef (VarDef name t) = do
+    transformedT <- transformXType @a @b t
+    pure $ VarDef name transformedT
+
+transformLocator :: forall a b c. (Phase a, Phase b, Phase c, ProgramTransform a b, PhaseTransition a c) => Locator a -> State (Env c) (Locator b)
+transformLocator (IntConst i) = pure $ IntConst i
+transformLocator (BoolConst b) = pure $ BoolConst b
+transformLocator (StrConst s) = pure $ StrConst s
+transformLocator (AddrConst addr) = pure $ AddrConst addr
+transformLocator (Var var) = pure $ Var var
+transformLocator (Field l name) = do
+    transformedL <- transformLocator l
+    pure $ Field transformedL name
+
+transformLocator (Multiset t locators) = do
+    transformedT <- transformXType @a @b t
+    transformedLocators <- mapM transformLocator locators
+    pure $ Multiset transformedT transformedLocators
+
+transformLocator (NewVar name baseT) = do
+    transformedBaseT <- transformBaseType baseT
+    pure $ NewVar name transformedBaseT
+
+transformLocator Consume = pure Consume
+
+transformLocator (RecordLit keys members) = do
+    let (varDefs, locators) = unzip members
+    transformedVarDefs <- mapM transformVarDef varDefs
+    transformedLocators <- mapM transformLocator locators
+    pure $ RecordLit keys (zip transformedVarDefs transformedLocators)
+
+transformLocator (Filter l q predName args) = do
+    transformedL <- transformLocator l
+    transformedArgs <- mapM transformLocator args
+    pure $ Filter transformedL q predName transformedArgs
+
+transformLocator (Select l k) = do
```

Comment:
A lot of these could be simplified using the `Applicative` operators into something like `Select <$> transformLocator l <*> transformLocator k`